src/components/molecules/ChatPoll/ChatPoll.tsx

```diff
@@ -0,0 +1,131 @@
+import React, { useCallback, useMemo } from "react";
+import classNames from "classnames";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { faPoll } from "@fortawesome/free-solid-svg-icons";
+
+import { PollMessage, BaseMessageToDisplay, PollValues } from "types/chat";
+
+import { WithId } from "utils/id";
+
+import { ChatMessageInfo } from "components/atoms/ChatMessageInfo";
+import { TextButton } from "components/atoms/TextButton";
+import Button from "components/atoms/Button";
+
+import { useUser } from "hooks/useUser";
+
+import "./ChatPoll.scss";
+
+export interface ChatPollProps {
+  pollData: WithId<BaseMessageToDisplay<PollMessage>>;
+  deletePoll: (pollId: string) => void;
+  voteInPoll: (poll: PollValues, votes: string[], pollId: string) => void;
+}
+
+export const ChatPoll: React.FC<ChatPollProps> = ({
+  pollData,
+  voteInPoll,
+  deletePoll,
+}) => {
+  const { user } = useUser();
+
+  const userId = user?.uid;
+
+  const { id, poll, votes, isMine, canBeDeleted } = pollData;
+  const { questions, topic } = poll;
+
+  const isVoted = userId ? votes.includes(userId) : false;
+
+  const containerStyles = classNames("ChatPoll", {
+    "ChatPoll--me": isMine,
+  });
+
+  const formattedQuestions = questions
+    .map((item) => ({
+      ...item,
+      count: votes.length
+        ? Math.floor(((item.votes || 0) / votes.length) * 100)
+        : 0,
+    }))
+    .sort((a, b) => b.count - a.count);
+
+  const handleVote = useCallback(
+    (question) => {
+      if (!userId) return;
+
+      const newVotes = [...votes, userId];
+
+      const newPoll = {
+        topic,
+        questions: questions.map((q) =>
+          q.id === question.id
+            ? {
+                ...question,
+                // think about to save all polling userIds
+                votes: question.votes + 1,
+              }
+            : q
+        ),
+      };
+      voteInPoll(newPoll, newVotes, id);
```

Comment:
these calculations/new object formations must happen in a secure place (backend fn)

Reply:
Instead of saving a count of votes, it’s probably better to just store an array of userIds that have voted. Then we can use the FieldValue.arrayUnion when users try to vote to ensure they can only ever vote once.

Reply:
fixed in 4e7f06c35c6b2ca0f88795306c915c2bf8143e3a

Reply:
The changes in that commit to the backend function don't look like they are using `FieldValue.arrayUnion`/similar?

How is this intended to be working currently based on the code here?

We're also not validating the content of `votes` that is passed to the backend function before passing it straight to `.update`, which is most likely a security risk.